### ISSUE TITLE
Crash if the battle intro is still playing but the user has already won the battle

### ISSUE THREAD

**Issue by gestern** [vcmi/vcmi#3172]
**Describe the bug**
The game crashes if the battle music is still playing but the user has already won the battle (assuming you go first and the player has a weak and slow unit). The problem **may not be reproduced every time**. When studying, approximately, the problem occurs somewhere every 3 times

**Game logs**
`VCMI_client.txt`:
```
[2023-Nov-11 03:08:00.342674] INFO global - Opening BattleAI
[2023-Nov-11 03:08:00.342839] INFO global - Loaded Battle AI
[2023-Nov-11 03:08:00.343011] DEBUG global - Building configurable interface object
[2023-Nov-11 03:08:00.343023] DEBUG global - Read variable named actionIconAttack
[2023-Nov-11 03:08:00.343030] DEBUG global - Read variable named actionIconDefault
[2023-Nov-11 03:08:00.343037] DEBUG global - Read variable named actionIconNoReturn
[2023-Nov-11 03:08:00.343044] DEBUG global - Read variable named actionIconReturn
[2023-Nov-11 03:08:00.343050] DEBUG global - Read variable named actionIconShoot
[2023-Nov-11 03:08:00.343057] DEBUG global - Read variable named actionIconSpell
[2023-Nov-11 03:08:00.343063] DEBUG global - Read variable named actionIconWalk
[2023-Nov-11 03:08:00.343070] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.343076] DEBUG global - Building widget CPicture
[2023-Nov-11 03:08:00.343082] DEBUG global - Reading point
[2023-Nov-11 03:08:00.343363] DEBUG global - Building widget with name menuTactics
[2023-Nov-11 03:08:00.343372] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.343378] DEBUG global - Building widget CPicture
[2023-Nov-11 03:08:00.343383] DEBUG global - Reading point
[2023-Nov-11 03:08:00.343901] DEBUG global - Building widget with name menuBattle
[2023-Nov-11 03:08:00.343912] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.343918] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.343941] DEBUG global - Reading point
[2023-Nov-11 03:08:00.343952] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.343964] DEBUG global - Reading hint text (help) from generaltext handler:core.help.381d
[2023-Nov-11 03:08:00.343991] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344006] DEBUG global - Building widget with name options
[2023-Nov-11 03:08:00.344015] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344021] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344029] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344035] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344041] DEBUG global - Reading hint text (help) from generaltext handler:core.help.379d
[2023-Nov-11 03:08:00.344058] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344067] DEBUG global - Building widget with name surrender
[2023-Nov-11 03:08:00.344074] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344079] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344085] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344090] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344096] DEBUG global - Reading hint text (help) from generaltext handler:core.help.380d
[2023-Nov-11 03:08:00.344114] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344125] DEBUG global - Building widget with name flee
[2023-Nov-11 03:08:00.344131] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344137] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344142] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344147] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344153] DEBUG global - Reading hint text (help) from generaltext handler:core.help.382d
[2023-Nov-11 03:08:00.344167] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344175] DEBUG global - Building widget with name autofight
[2023-Nov-11 03:08:00.344182] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344187] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344192] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344197] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344203] DEBUG global - Reading hint text (help) from generaltext handler:core.help.385d
[2023-Nov-11 03:08:00.344222] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344229] DEBUG global - Building widget with name cast
[2023-Nov-11 03:08:00.344235] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344240] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344245] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344250] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344256] DEBUG global - Reading hint text (help) from generaltext handler:core.help.386d
[2023-Nov-11 03:08:00.344270] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344277] DEBUG global - Building widget with name wait
[2023-Nov-11 03:08:00.344283] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344289] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344294] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344299] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344305] DEBUG global - Reading hint text (help) from generaltext handler:core.help.387d
[2023-Nov-11 03:08:00.344320] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344327] DEBUG global - Building widget with name defence
[2023-Nov-11 03:08:00.344333] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344339] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344344] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344349] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344362] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344369] DEBUG global - Building widget with name consoleUp
[2023-Nov-11 03:08:00.344376] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344381] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344386] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344391] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344402] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344409] DEBUG global - Building widget with name consoleDown
[2023-Nov-11 03:08:00.344415] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344421] DEBUG global - Reading rect
[2023-Nov-11 03:08:00.344427] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344434] DEBUG global - Building widget with name console
[2023-Nov-11 03:08:00.344440] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344445] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344450] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344455] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344470] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344477] DEBUG global - Building widget with name tacticNext
[2023-Nov-11 03:08:00.344484] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344489] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344497] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344502] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344519] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344526] DEBUG global - Building widget with name tacticEnd
[2023-Nov-11 03:08:00.344532] DEBUG global - Building widget from config
[2023-Nov-11 03:08:00.344537] DEBUG global - Building widget CButton
[2023-Nov-11 03:08:00.344543] DEBUG global - Reading point
[2023-Nov-11 03:08:00.344548] DEBUG global - Reading hint text
[2023-Nov-11 03:08:00.344861] DEBUG global - Reading hotkey
[2023-Nov-11 03:08:00.344872] DEBUG global - Building widget with name alternativeAction
[2023-Nov-11 03:08:00.393192] DEBUG global - totalRedraw requested!
[2023-Nov-11 03:08:00.400318] DEBUG animation - Activating next stack
```

`VCMI_server.txt`:
```
[2023-Nov-11 03:08:00.324585] DEBUG global - Hero hero.core.grindan.name visits Малком Друид (34:28)
[2023-Nov-11 03:08:00.326385] DEBUG global - Round -1
[2023-Nov-11 03:08:01.440700] ERROR network - Network error receiving a pack. Connection has been closed
[2023-Nov-11 03:08:01.440834] INFO network - Thread listening for Connection with VCMI 1.3.2 (server) (ID: 1 UUID: 1867eb67-010d-40e6-851c-48a13b8dd6da) ended
```

Coming from the stack (see below) the problem occurs when calling the method `BattleInterface::onIntroSoundPlayed()`:
https://github.com/vcmi/vcmi/blob/439ce114cbc7322f258bcc804eaf23b7fe50131f/client/battle/BattleInterface.cpp#L132

 **Call Stack:**
```
Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Termination Reason:    Namespace SIGNAL, Code 6 Abort trap: 6
Terminating Process:   vcmiclient [13809]

Application Specific Information:
abort() called


Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	       0x188190764 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x1881c7c28 pthread_kill + 288
2   libsystem_c.dylib             	       0x1880d5ae8 abort + 180
3   libc++abi.dylib               	       0x188180b84 abort_message + 132
4   libc++abi.dylib               	       0x1881703b4 demangling_terminate_handler() + 320
5   libobjc.A.dylib               	       0x187e4703c _objc_terminate() + 160
6   libc++abi.dylib               	       0x18817ff48 std::__terminate(void (*)()) + 16
7   libc++abi.dylib               	       0x188182d34 __cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) + 36
8   libc++abi.dylib               	       0x188182ce0 __cxa_throw + 140
9   vcmiclient                    	       0x1047c5548 void boost::throw_exception<boost::lock_error>(boost::lock_error const&) + 60
10  vcmiclient                    	       0x1047c53dc boost::unique_lock<boost::mutex>::lock() + 212
11  vcmiclient                    	       0x104839788 CondSh<bool>::setn(bool) + 44
12  vcmiclient                    	       0x104835ee4 BattleInterface::onIntroSoundPlayed() + 36
13  vcmiclient                    	       0x10483a258 std::__1::__function::__func<BattleInterface::playIntroSoundAndUnlockInterface()::$_0, std::__1::allocator<BattleInterface::playIntroSoundAndUnlockInterface()::$_0>, void ()>::operator()() + 72
14  vcmiclient                    	       0x104abd29c std::__1::__function::__func<CSoundHandler::soundFinishedCallback(int)::$_2, std::__1::allocator<CSoundHandler::soundFinishedCallback(int)::$_2>, void ()>::operator()() + 160
15  vcmiclient                    	       0x1048784f0 InputHandler::fetchEvents() + 64
16  vcmiclient                    	       0x104aa98cc main + 17472
17  dyld                          	       0x187e6ff28 start + 2236
```

**To Reproduce**
1. Start a battle with an opponent who has one weak unit;
2. Without waiting for the intro music to end, kill your opponent's only unit;
3. A crash MAY occur.

**Expected behavior**
The intro music should stop and the battle victory music should start playing.

**Actual behavior**
The game crashes if the battle music is still playing but the user has already won the battle

**Did it work earlier?**
n/a

**Screenshots**
<img width="1424" alt="image" src="https://github.com/vcmi/vcmi/assets/1861589/fa8080e9-4082-49b1-b77d-ffb4d1697be4">

**Version**
 - OS: `macOS ARM 13.5`
 - Version: `VCMI 1.3.2`

**Additional context**
n/a

**Comment by nullkiller** [comment]
Was not able to reproduce on Windows. Battle into sound cracks a bit but still no crash.

**Comment by IvanSavenko** [comment]
Closing, should be fixed in 1.4
